### PR TITLE
add fsdp checkpoint support for custom device

### LINKED ISSUE #104390: fsdp checkpoint wants to support custom device
### 🐛 Describe the bug

As for fsdp checkpoint, there are some hard-code only for `cuda`, for custom device (privateuse1 backend, cuda-like device), we also want to support it.

### Versions

main

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes https://github.com/pytorch/pytorch/issues/104390
Now for distributed checkpointing, there are some hard-code like cuda, we want to support checkpointing for custom device( privateuse1 backend), so we use the abstract device-module to run some funcs.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by heidongxianhua** [comment]
@awgu  Hi sorry to bother, could you have a look? we want to support fsdp related functions for custom device(privateuse1 backend), so we give a few PRs to do it, and another PR is https://github.com/pytorch/pytorch/pull/104155, it is to support the distributed hooks for custom device.

**Comment by heidongxianhua** [comment]
could you have a look at this pr and #104155 @fegin @wz337 ？

File: torch/distributed/utils.py
Comment (kumpera) [inline torch/distributed/utils.py:334]: @awgu FSDP has code that is almost identical to this one.

Can we unify it all here?

File: torch/distributed/checkpoint/filesystem.py
Comment (kumpera) [inline torch/distributed/checkpoint/filesystem.py:131]: I'm trying to make the "cpu" device have the same API here: https://github.com/pytorch/pytorch/pull/103172

Once that change is merged, we can remove all those asserts and conditional checks for stream device.
I'd love to move all the code we're adding in utils.py to a top level module that would even include protocols that each device type needs to provide.

File: torch/distributed/utils.py
Comment (awgu) [reply torch/distributed/utils.py:334]: I might have missed it -- where in FSDP code do we convert from string (e.g. `"cuda"`) to the Python module (e.g. `torch.cuda`)?

The closest thing I can see is the function `_init_device_handle()` with:
https://github.com/pytorch/pytorch/blob/4baac20117a86f811aa9acf5f8d9d3415faff238/torch/distributed/fsdp/_init_utils.py#L348

**Comment by heidongxianhua** [comment]
> > could you have a look at this pr and #104155 @fegin @wz337 ？
> 
> @heidongxianhua Thanks for contributing. The change looks good to me.
> 
> Curious, did you happen to add a test on fsdp side for custom device? I wonder whether you could add a test for this.

ehhah, yeah, we are working to make custom device support fsdp at this time, and the testcase is going on, and it is a bit different to  do this.

File: torch/distributed/checkpoint/filesystem.py
Comment (heidongxianhua) [reply torch/distributed/checkpoint/filesystem.py:131]: > 

yes, that change looks cool. It  will make it easy to extend new device, we can process the `cpu` device same as `cuda` and `privateuse1` backend. It is also very useful for other modules.  And I will unify the code once your change is merged. @kumpera

Comment (heidongxianhua) [reply torch/distributed/checkpoint/filesystem.py:131]: And another idea, maybe we could to this APIs with the same func but with a device argument, like `torch.stream(device="cuda")` for `cuda`, and `torch.stream(device="cpu")` for `cpu`. And it can unify the APIs, but it will make a big change. I do not know which is better now. @kumpera

**Comment by heidongxianhua** [comment]
@kumpera  I have move the `_get_device_module` code to the top  level `utils.py` so that we could reuse it in other modules. And could you have a review again? and the another PR https://github.com/pytorch/pytorch/pull/104155 is also waiting for review.

File: torch/distributed/checkpoint/_fsspec_filesystem.py
Comment (kumpera) [inline torch/distributed/checkpoint/_fsspec_filesystem.py:300]: nit: maybe change it to `assert tensor.is_cpu` as that's what we need to ensure.

File: torch/distributed/checkpoint/_fsspec_filesystem.py
Comment (kumpera) [inline torch/distributed/checkpoint/_fsspec_filesystem.py:118]: The problem of using torch.Stream is, AFAICT, it won't typecheck `torch.cuda.Stream` as the later doesn't inherit from the former. We likely need to use a protocol instead.

File: torch/distributed/checkpoint/_fsspec_filesystem.py
Comment (kumpera) [inline torch/distributed/checkpoint/_fsspec_filesystem.py:128]: Shouldn't this be `stream.device if stream else torch.device("cuda")` ?

File: torch/distributed/checkpoint/_sharded_tensor_utils.py
Comment (kumpera) [inline torch/distributed/checkpoint/_sharded_tensor_utils.py:95]: This is wrong, it won't work with CPU

File: torch/distributed/checkpoint/optimizer.py
Comment (kumpera) [inline torch/distributed/checkpoint/optimizer.py:55]: This is wrong as it will produce `cpu:0` for `cpu`.

We need to add a helper to `torch.cuda` and `torch.cpu` to format a number into a device string.

**Review by kumpera — CHANGES_REQUESTED** [review]

File: torch/distributed/checkpoint/_fsspec_filesystem.py
Comment (heidongxianhua) [reply torch/distributed/checkpoint/_fsspec_filesystem.py:118]: > The problem of using torch.Stream is, AFAICT, it won't typecheck `torch.cuda.Stream` as the later doesn't inherit from the former. We likely need to use a protocol instead.

ehah, in fact, the `torch.cuda.Stream` is inherited from `torch.Stream`, the relationship between two class is as follows:
image

And I have tested it, 
```
>>> import torch
>>> issubclass(torch.cuda.Stream, torch.Stream)
True
>>> torch.__version__
'2.1.0a0+git1204463'
```

So maybe it is ok to use `torch.Stream`? @kumpera

**Comment by heidongxianhua** [comment]
@kumpera could have a review again?

**Pull request closed without merging**
